src/internal/layout-engine/engine-solution.ts

```diff
@@ -0,0 +1,251 @@
+      score,
+    };
+  }
+}
+
+/**
+ * Given a solution state finds a set of all possible moves each resolving a particular overlap.
+ */
+export function findNextSolutions(state: MoveSolutionState): MoveSolution[] {
+  // For every overlap and direction found a move if exists that resolves the overlap.
+  // A pair of the given state and the overlap resolution move is a new solution to try.
+  const nextMoveSolutions: MoveSolution[] = [];
+  for (const [overlapId, overlapIssuerId] of state.overlaps) {
+    for (const moveDirection of PRIORITY_DIRECTIONS) {
+      const move = getOverlapMove(state, overlapId, overlapIssuerId, moveDirection);
+      if (move !== null) {
+        nextMoveSolutions.push([MoveSolutionState.clone(state), move]);
+      }
+    }
+  }
+  return nextMoveSolutions;
+}
+
+// Returns an evaluated move to resolve the given overlap in the given direction or null if such move is not possible.
+function getOverlapMove(
+  state: MoveSolutionState,
+  overlapId: ItemId,
+  overlapIssuerId: ItemId,
+  moveDirection: Direction
+): null | CommittedMove {
+  const userItem = state.grid.getItem(state.moves[0].itemId);
+  const overlapItem = state.grid.getItem(overlapId);
+  const overlapIssuerItem = state.grid.getItem(overlapIssuerId);
+  const overlapMove = getMoveForDirection(overlapItem, overlapIssuerItem, moveDirection);
+
+  // The move position is outside the grid boundaries.
+  if (overlapMove.x < 0 || overlapMove.y < 0 || overlapMove.x + overlapMove.width > state.grid.width) {
+    return null;
+  }
+
+  // Subsequent item overlap moves in the opposite directions do not contribute to solution.
+  const prevOverlapMove = getLastSolutionMove(state, overlapItem.id);
+  if (prevOverlapMove && checkOppositeDirections(prevOverlapMove.direction, moveDirection)) {
+    return null;
+  }
+
+  const pathOverlaps = getPathOverlaps(state, overlapMove, overlapIssuerItem);
+  for (const overlap of pathOverlaps) {
+    // Not allowed to intersect with the user-controlled item.
+    if (overlap.id === userItem.id) {
+      return null;
+    }
+    // Not allowed to intersect with conflicting items.
+    if (state.conflicts?.items.has(overlap.id)) {
+      return null;
+    }
+    // Intersecting with items having unresolved overlaps does not contribute to solution.
+    if (state.overlaps.has(overlap.id)) {
+      return null;
+    }
+  }
+
+  const lastIssuerMove = getLastSolutionMove(state, overlapIssuerItem.id);
+  if (!lastIssuerMove) {
+    throw new Error("Invariant violation: overlap issuer has no associated moves.");
+  }
+  const issuerDirection = lastIssuerMove.direction;
+
+  const isSwap = checkIfSwap(overlapMove, lastIssuerMove);
+  const isDifferentIssuerDirection = moveDirection !== issuerDirection;
+  const isOppositeIssuerDirection = checkOppositeDirections(moveDirection, issuerDirection);
+  const userMoveBoundaries = getUserMoveBoundaries(state);
+  const moveVector = getSolutionMovesVector(state);
```

Comment:
The vector represents overlap resolutions made before. If we vector would include the overlap issuer it can probably be merged with isDifferentIssuerDirection/isOppositeIssuerDirection properties. The idea is similar - the overlaps should be resolved with a common direction vector. The moves in the opposite directions do not typically contribute to a solution and can be unexpected from the user perspective.